ready for review (#51)
rstudio/tensorflow-blog: _posts/2019-10-01-intro-to-hmc/hmc-intro.Rmd

---
title: "On leapfrogs, crashing satellites, and going nuts: A very first conceptual introduction to Hamiltonian Monte Carlo"
description: > 

>   TensorFlow Probability, and its R wrapper tfprobability, provide Markov Chain Monte Carlo (MCMC) methods that were used in a number of recent posts on this blog. These posts were directed to users already comfortable with the method, and terminology, per se, which readers mainly interested in deep learning won't necessarily be. Here we try to make up leeway, introducing Hamitonian Monte Carlo (HMC) as well as a few often-heard "buzzwords" accompanying it, always striving to keep in mind what it is all "for".

Reviewer: > which readers mainly interested in deep learning won't necessarily be

I feel like this is sort of implied, or it could be part of the following line introducing the audience for and what you will do in this post.

author:
  - name: Sigrid Keydana
    affiliation: RStudio
    affiliation_url: https://www.rstudio.com/
slug: keydana2019hmc
date: 10-01-2019
categories:
  - Probability and statistics
  - Concepts
  - Introductions
bibliography: bibliography.bib
output:
  distill::distill_article:
    self_contained: false
preview: images/mb.png
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```

Why a _very (meaning: VERY!) first conceptual introduction to Hamiltonian Monte Carlo_ (HMC) on this blog? 

> Well, in our endeavor to feature the various capabilities of TensorFlow Probability (TFP) / [tfprobability](https://rstudio.github.io/tfprobability/), we started showing examples of how to fit hierarchical models, using one of TFP's joint distribution classes ^[[tfd_joint_distribution_sequential](https://rstudio.github.io/tfprobability/reference/tfd_joint_distribution_sequential.html)] and HMC. The technical aspects being complex enough in themselves, we never provided an introduction to the "math side of things". For those of our readers who are familiar with the technique (Stan users, for example), we hope this was acceptable; but how about those who are following the blog -- the _TensorFlow for R_ blog, alright -- due to their interest in _deep learning_? They may have felt left behind or -- put more prosaically -- just have stopped reading instantaneously.

Reviewer: > The technical aspects being complex enough in themselves

You might phrase this so it sounds more like a qualifier (the construct: "The technical aspects being..." just sounds a little off to my ear). 

Perhaps something like: Given the complexity of the technical implementation, we never provided...

This post is directed towards the second group of readers, aiming to make up for the imposition.

> If after reading, you still feel confused, it isn't your fault and _maybe_, not even the fault of this text. Based on an inofficial, minimal-sample-size survey, this topic is a _lot_ more confusing than deep learning :-). As to the content, we'll try to keep it as digestable as possible, while at the same time striving to mention all of the famous incantations: Hamiltonian Monte Carlo, _leapfrog_ steps, NUTS.

Reviewer: ```suggestion
If after reading, you still feel confused, it isn't your fault and _maybe_, not even the fault of this text. Based on an unofficial, minimal-sample-size survey, this topic is a _lot_ more confusing than deep learning :-). As to the content, we'll try to keep it as digestible as possible, while at the same time striving to mention all of the famous incantations: Hamiltonian Monte Carlo, _leapfrog_ steps, NUTS.
```

> ## So what is it for?

Reviewer: Is the "it" here HMC? If so, you might replace the "it" with HMC.

> Sampling, or _Monte Carlo_, techniques in general are used when we want to produce samples from, or statistically describe, a distribution we don't have a closed-form formulation of. Sometimes, we might really be interested in the samples; sometimes we just want them so we can compute, for example, the mean and variance of the distribution.

Reviewer: ```suggestion
Sampling, or _Monte Carlo_, techniques in general are used when we want to produce samples from, or statistically describe a distribution we don't have a closed-form formulation of. Sometimes, we might really be interested in the samples; sometimes we just want them so we can compute, for example, the mean and variance of the distribution.
```

> What distribution? In the type of applications we're talking about, we have a _model_, a joint distribution, supposed to describe some reality. Starting from the most basic scenario, it might look like this:

Reviewer: ```suggestion
What distribution? In the type of applications we're talking about, we have a _model_, a joint distribution, which is supposed to describe some reality. Starting from the most basic scenario, it might look like this:
```

$$
x \sim \mathcal{Poisson}(\lambda)
$$

> This "joint distribution" only has a single member, a Poisson distribution, that is supposed to model, say, the number of comments in a code review. We also have data on actual code reviews, like this, say ^[data are purely made up]:

Reviewer: The first line has quite a few asides. Given the heavy use of jargon and the intended audience, you might break it into two. In the first use a generic term (or none at all). And, in the latter, you could give the example ("the number of comments in a code review"). 

The other option would be to introduce the example you'll be discussing first, so that there's something for the reader to grab onto mentally/it's not buried later in the paragraph.

```{r, echo = FALSE}
ggplot(data.frame(x = x), aes(x = x)) + geom_bar() + ggtitle("Number of comments in (fictional) code reviews") + labs(x = "number of comments", y = "number of code reviews") + scale_x_discrete(drop=FALSE)
```

```{r, eval=TRUE, echo=FALSE, layout="l-body", fig.cap = ""}
knitr::include_graphics("images/comments.png")
```

> We now want to determine the _parameter_, $\lambda$, of the Poisson that make these data most _likely_. So far, we're not even beign Bayesian yet: There is no prior on this parameter. But of course, we want to be Bayesian, so we add one -- imagine fixed priors ^[elided] on _its_ parameters:

Reviewer: ```suggestion
We now want to determine the _parameter_, $\lambda$, of the Poisson that make these data most _likely_. So far, we're not even being Bayesian yet: There is no prior on this parameter. But of course, we want to be Bayesian, so we add one -- imagine fixed priors ^[elided] on _its_ parameters:
```

$$
x \sim \mathcal{Poisson}(\lambda)\\
\lambda \sim \gamma(\alpha, \beta)\\
\alpha \sim [...]\\  
\beta \sim [...]
$$

> This now really being a joint distribution, we have three parameters to determine: $\lambda$, $\alpha$ and $\beta$.

Reviewer: ```suggestion
This being a joint distribution, we have three parameters to determine: $\lambda$, $\alpha$ and $\beta$.
```

Reviewer: If there's a concrete way to hint at what lambda, alpha, and beta represent, you might consider adding that. (I don't have a suggestion, because I'm learning this as I read it 😉).

Author: I'll do that. The thing is, I deliberately didn't want to go into too much detail here... but let me check if I can find a compromise. See also my general answer upcoming...

And what we're interested in is the _posterior distribution_ of the parameters given the data.

> Now many expositions will tell you how, depending on the distributions involved, we usually cannot calculate the posterior distributions in closed form. Instead of reiterating these explanations, here we'll just build on them and state that we have to use sampling techniques to determine those parameters. ^[In some cases, variational inference is an alternative. See [@Blei_2017] for a nice introduction.] What we'd like to point out instead is the following: In the upcoming discussions of sampling, HMC & co., it is really easy to forget _what is it that we are sampling_. Try to always keep in mind that what we're sampling isn't the data, it's parameters: the parameters of the posterior distributions we're interested in.

Reviewer: > Now many expositions will tell you how, depending on the distributions involved, we usually cannot calculate the posterior distributions in closed form.

Could this be reduced to something like: Usually, we cannot calculate the posterior distributions in closed form. ?

Then you have the ~ "see a bunch of other things if you want to understand this" afterward.

In Neal's own words

> In two dimensions, we can visualize the dynamics as that of a frictionless puck that slides over a surface of varying height. The state of this system consists of the position of the puck, given by a 2D vector q, and the momentum of the puck (its mass times its velocity), given by a 2D vector p.

Now when you hear "momentum" (and given that I've primed you to think of deep learning) you may feel that sounds familiar, but even though the respective analogies are related the association does not help that much. In deep learning, momentum is commonly praised for its avoidance of ineffective oscillations in imbalanced optimization landscapes ^[Though see [this great distill.pub post](https://distill.pub/2017/momentum/) for a different view.].
With HMC however, the focus is on the concept of _energy_.

> In [statistical mechanics](https://en.wikipedia.org/wiki/Statistical_mechanics), the probability of being in some state is inverse-exponentially related to its energy: ^[Here $T$ is the _temperature_; we won't focus on this so just imagine it being set to 1 in this and subsequent equations.]

Reviewer 2: ```suggestion
In [statistical mechanics](https://en.wikipedia.org/wiki/Statistical_mechanics), the probability of being in some state  $i$ is inverse-exponentially related to its energy: ^[Here $T$ is the _temperature_; we won't focus on this so just imagine it being set to 1 in this and subsequent equations.]
```

Also consider bringing the footnote out into the main text in parentheses? I missed it the first time and wondered what T was.

Author: Makes sense, will do!

Author: GH didn't let me commit the suggestion, fixed manually

  $$P(E_i) \sim e^{\frac{-E_i}{T}} $$

> As you might or might not remember from school physics, energy comes in two forms: potential energy and kinetic energy. In the sliding-object scenario, the object's potential energy corresponds to its height (position), while its kinetic energy is related to its momentum by the formula ^[$m$ here stands for momentum.]

Reviewer 2: Also is writing out 'mass' standard notation?

> As you might or might not remember from school physics, energy comes in two forms: potential energy and kinetic energy. In the sliding-object scenario, the object's potential energy corresponds to its height (position), while its kinetic energy is related to its momentum by the formula ^[$m$ here stands for momentum.]

Author: I don't think so, probably not... I was trying to exactly follow the notation(s) of my sources, (1) the Ben Lambert video and (2) Radford Neal's article.
I'm also trying to say that explicitly (in case someone wonders why the notation changes).
In the beginning, I was thinking to "unify" (wrong word but you know what I mean) it all, but in the end it (1) seemed safest and (2) should be nice for readers who go on to watch/read those original sources directly...
Does this make sense?

> As you might or might not remember from school physics, energy comes in two forms: potential energy and kinetic energy. In the sliding-object scenario, the object's potential energy corresponds to its height (position), while its kinetic energy is related to its momentum by the formula ^[$m$ here stands for momentum.]

Reviewer 2: yep if that's what the references use it's fine

> As you might or might not remember from school physics, energy comes in two forms: potential energy and kinetic energy. In the sliding-object scenario, the object's potential energy corresponds to its height (position), while its kinetic energy is related to its momentum by the formula ^[$m$ here stands for momentum.]

Author: that's good to hear!

$$U(\theta) \sim - log (P(x | \theta) P(\theta))$$

So the "trick" of HMC is augmenting the state space of interest - the vector of posterior parameters - by a momentum vector, to improve optimization efficiency. When we're finished, the momentum part is just thrown away. (This aspect is especially nicely explained in Ben Lambert's video.)

Following his exposition and notation, here we have the energy of a state of parameter and momentum vectors, equaling a sum of potential and kinetic energies:

$$E(\theta, m) = U(\theta) + K(m)$$

The corresponding probability, as per the relationship given above, then is

$$P(E) \sim e^{\frac{-E}{T}} = e^{\frac{- U(\theta)}{T}} e^{\frac{- K(m)}{T}}$$ 

We now substitute into this equation, assuming a temperature (T) of 1 and a mass of 1:

$$P(E) \sim P(x | \theta) P(\theta) e^{\frac{- m^2}{2}}$$

Now in this formulation, the distribution of momentum is just a standard normal ($e^{\frac{- m^2}{2}}$)! Thus, we can just integrate out the momentum and take $P(\theta)$ as samples from the posterior distribution:^[Here $Z$ is the normalizer required to have the integral sum to 1.]

$$

> \begin{aligned}

Reviewer 2: I'd begin this equation with P(\theta) = \int P(\theta, m) dm = ...

---
title: "On leapfrogs, crashing satellites, and going nuts: A very first conceptual introduction to Hamiltonian Monte Carlo"
description: > 
  TensorFlow Probability, and its R wrapper tfprobability, provide Markov Chain Monte Carlo (MCMC) methods that were used in a number of recent posts on this blog. These posts were directed to users already comfortable with the method, and terminology, per se, which readers mainly interested in deep learning won't necessarily be. Here we try to make up leeway, introducing Hamitonian Monte Carlo (HMC) as well as a few often-heard "buzzwords" accompanying it, always striving to keep in mind what it is all "for".
author:
  - name: Sigrid Keydana
    affiliation: RStudio
    affiliation_url: https://www.rstudio.com/
slug: keydana2019hmc
date: 10-01-2019
categories:
  - Probability and statistics
  - Concepts
  - Introductions
bibliography: bibliography.bib
output:
  distill::distill_article:
    self_contained: false
preview: images/mb.png
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```

Why a _very (meaning: VERY!) first conceptual introduction to Hamiltonian Monte Carlo_ (HMC) on this blog? 

> Well, in our endeavor to feature the various capabilities of TensorFlow Probability (TFP) / [tfprobability](https://rstudio.github.io/tfprobability/), we started showing examples of how to fit hierarchical models, using one of TFP's joint distribution classes ^[[tfd_joint_distribution_sequential](https://rstudio.github.io/tfprobability/reference/tfd_joint_distribution_sequential.html)] and HMC. The technical aspects being complex enough in themselves, we never gave an introduction to the "math side of things". Here we are trying to make up for this. 

Reviewer 3: Should we add references to https://blogs.rstudio.com/tensorflow/posts/2019-06-25-dynamic_linear_models_tfprobability/ and https://blogs.rstudio.com/tensorflow/posts/2019-05-06-tadpoles-on-tensorflow/ after HMC? It would help for readers that want to jump to the application of HMC before understanding how it works.

Author: good idea, i'll add them in a footnote!

Now in this formulation, the distribution of momentum is just a standard normal ($e^{\frac{- m^2}{2}}$)! Thus, we can just integrate out the momentum and take $P(\theta)$ as samples from the posterior distribution:^[Here $Z$ is the normalizer required to have the integral sum to 1.]

$$
\begin{aligned}
& P(\theta) = 
\int \! P(\theta, m) \mathrm{d}m = \frac{1}{Z} \int \! P(x | \theta) P(\theta) \mathcal{N}(m|0,1) \mathrm{d}m\\
& P(\theta) = \frac{1}{Z} \int \! P(x | \theta) P(\theta)
\end{aligned}
$$

How does this work in practice? At every step, we

- sample a new momentum value from its marginal distribution (which is the same as the conditional distribution given $U$, as they are independent), and
- solve for the path of the particle. This is where _Hamilton's equations_ come into play.

## Hamilton's equations (equations of motion)

For the sake of less confusion, should you decide to read the paper, here we switch to Radford Neal's notation. 

Hamiltonian dynamics operates on a d-dimensional position vector, $q$, and a d-dimensional momentum vector, $p$. The state space is described by the _Hamiltonian_, a function of $p$ and $q$:

$$H(q, p) =U(q) +K(p)$$

> Here $U(q)$ is the potential energy (called $U(\theta)$ above), and $K(p)$ is the potential energy as a function of momentum (called $K(m)$ above).

Reviewer 2: ```suggestion
Here $U(q)$ is the potential energy (called $U(\theta)$ above), and $K(p)$ is the kinetic energy as a function of momentum (called $K(m)$ above).
```

NUTS being more of algorithm in the computer science usage of the word than a thing to explain conceptually, we'll leave it at that, and ask the interested reader to read the paper -- or even, [consult the TFP documentation to see how NUTS is implemented there](https://github.com/tensorflow/probability/blob/master/discussion/technical_note_on_unrolled_nuts.md). Instead, we'll round up with another conceptual analogy, Michael Bétancourts crashing (or not!) satellite [@2017arXiv170102434B].

## How to avoid crashes

Bétancourt's article is an awesome read, and a paragraph focusing on a single point made in the paper can be nothing than a "teaser" (which is why we'll have a picture, too!). 

To introduce the upcoming analogy, the problem starts with high dimensionality, which is a given in most real-world problems. In high dimensions, as usual, the density function has a _mode_ (the place where it is maximal), but necessarily, there cannot be much _volume_ around it -- just like with k-nearest neighbors, the more dimensions you add, the farther your nearest neighbor will be.

> A product of volume and density, the only significant probability mass resides in the so-called _typical set_ ^[for a detailed discussion of typical sets, see the book by McKay [@MacKay:2002:ITI:971143].] , which becomes more and more narrow in high dimensions.

Reviewer 2: Might be worth adding a sentence or two on what typical sets are (if that's epistemically possible), since we're using the term down the post and people aren't likely going to acquire the book to read up on it

Author: yeah I see... I added a link to Wikipedia :-)
